### ISSUE TITLE
Loss function that accepts mask info

### ISSUE THREAD

**Issue by goksinan** [KevinMusgrave/pytorch-metric-learning#159]
As far as I understand, the loss functions operate only on the embeddings. Is it also possible to provide the corresponding "mask" information along with the embeddings so that the loss is computed using the "unmasked" region of the embeddings? Can you suggest a workaround for this?
Thank you!

**Comment by KevinMusgrave** [comment]
Do you mean masking out certain dimensions of each embedding vector? Can you give a simple example of what you're thinking of?

**Comment by goksinan** [comment]
That is correct. Let's say the the embedding is a vector of 1024 elements. Let's say the distance between embeddings _e1_ and _e2_ is `(e1 - e2).norm()`. Now, instead of using all 1024 of them, I want to mask out certain dimensions (from both vectors) that do not include any useful info, because I already know that the corresponding regions in the input images are distorted or blacked-out. Thus, I am trying to avoid including those regions in distance computation which can artificially increase the similarity between two images.

**Comment by KevinMusgrave** [comment]
If the batch size is N, will each masked embedding have the same size? For example, if the original size of ```e1``` and ```e2``` is 1024, and after being masked their size is 326, will other elements in the batch ```e3``` and ```e4``` also have size 326?

**Comment by goksinan** [comment]
No. Each embedding has its own mask and their sizes are different. While computing the distance, I superimpose the masks so that only the available dimensions on both vectors are used, such as: `d = (e1 - e2)[m1 AND m2].norm()` (_m1_ and _m2_ are mask vectors and boolean). I also normalize the distance by the number of available dimensions.

**Comment by KevinMusgrave** [comment]
So the masked e1 and e2 are the same size. But if other vectors in the batch (e3, e4, e5...) are a different size, it may be complicated to get it to work because a lot of the operations in the loss functions are batch operations, which expect all the vectors to be the same size.

**Comment by goksinan** [comment]
I understand. Thanks for your time.
